src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -1794,4 +1833,221 @@ public Vector mul(Vector point) {
             return new Vector(vx, vy, vz);
         }
     }
+
+    public interface ParticleData {
+        /**
+         * Returns true if the data type uses the offset fields,
+         * and false if it uses the data field.
+         * Sometimes depends on the particle and version.
+         */
+        public default boolean isOffsetBased(XParticle particle) {
+            return false;
+        }
+
+        public default double[] offsetValues() {
+            return new double[]{0, 0, 0};
+        }
```

Comment:
I think using a `Vector` here is more appropriate. Firstly because it's a fixed 3-valued object and because you wouldn't need to pass it around from spawn, spawnWithDataInOffset, and spawnRaw methods. An array would help the performance just a tiny bit, it's not worth it.

Reply:
Good point!